src/wasm/nxt_wasm.c

```diff
@@ -0,0 +1,262 @@
+/*
+ * Copyright (C) Andrew Clayton
+ * Copyright (C) F5, Inc.
+ */
+
+#include <nxt_main.h>
+#include <nxt_application.h>
+#include <nxt_unit.h>
+#include <nxt_unit_request.h>
+
+#include "nxt_wasm.h"
+
+#define NXT_WASM_VERSION    "0.1"
+
+static nxt_wasm_ctx_t               nxt_wasm_ctx;
+
+static const nxt_wasm_operations_t  *nxt_wops;
+
+#define NXT_WASM_DO_HOOK(hook)  nxt_wops->exec_hook(&nxt_wasm_ctx, hook);
+
+void
+nxt_wasm_do_response_end(nxt_wasm_ctx_t *ctx)
+{
+    nxt_unit_request_done(ctx->req, NXT_UNIT_OK);
+
+    NXT_WASM_DO_HOOK(NXT_WASM_FH_RESPONSE_END);
+}
+
+
+void
+nxt_wasm_do_send_headers(uint32_t offs, nxt_wasm_ctx_t *ctx)
+{
+    size_t                   fields_len;
+    unsigned int             i;
+    nxt_wasm_response_hdr_t  *rh;
+
+    rh = (nxt_wasm_response_hdr_t *)(ctx->baddr + offs);
+
+    fields_len = 0;
+    for (i = 0; i < rh->nr_fields; i++)
+        fields_len += rh->fields[i].name_len + rh->fields[i].value_len;
+
+    nxt_unit_response_init(ctx->req, 200, rh->nr_fields, fields_len);
+
+    for (i = 0; i < rh->nr_fields; i++) {
+        const char *name;
+        const char *val;
+
+        name = (const char *)(uint8_t *)rh + rh->fields[i].name_offs;
+        val = (const char *)(uint8_t *)rh + rh->fields[i].value_offs;
```

Comment:
Space after cast.

Reply:
BTW, why do you need the double cast?

Reply:
My comment here has vanished!

Reply:
I never received it :)

Reply:
It did need it at one point. Just checked again and now I only need

```C
        name = (const char *)rh + rh->fields[i].name_offs;
        val = (const char *)rh + rh->fields[i].value_offs;
```